### ISSUE TITLE
ajax() needs to include abort() in returned Promise

### ISSUE THREAD

**Issue by bmomberger-bitovi** [canjs/can-util#30]
the promise returned from `ajax()` should have `xhr.abort()` added to it, so the user can interrupt the XHR.

**Comment by prashantsharmain** [comment]
@bmomberger-bitovi @justinbmeyer 

This needs some discussion.

The ajax() call returns an ES 6 promise. We'll have to consider:
1. Do we want to alter Promise.prototype?
2. A promise cannot resolve or reject itself. Should it, then, be allowed to abort itself?

Thoughts?

**Comment by justinbmeyer** [comment]
@prashantsharmain I think in this case, we can "duck punch" an abort method on just these promises (modify the instance of promise instead of the Promise.prototype).

**Comment by bmomberger-bitovi** [comment]
The ajax() call doesn't necessarily have to return a proper-to-standard ES6 promise; we do that currently out of convenience.  Seconding the duck-punching of a Promise to something quacking like a jqXHR.

Key to your second point, @prashantsharmain, is that we're not intending to implement abort() with something akin to `var reject, p=Promise((_, rej) => {reject=rej}); assign(p, {abort:reject}); return p;`.  abort() won't take an argument with which to reject the promise, so we'll avoid the promise being rejected with itself (which anyway causes a TypeError in Chrome's Promise implementation as well as the Promise implementation in the `when` library we use for anything Steal-backed).

So there's no reason that I can see why we couldn't provide an "Abortable promise" that exposes an indirect rejection functionality.

**Comment by rjgotten** [comment]
You may want to think twice before implementing this as a duck-punched API extension on `Promise`, because it may set a precedent that will backfire on you in the future.

Worth reading:
https://esdiscuss.org/topic/cancelable-promises

Currently, consensus on task cancellation is being steered in the direction of an externally created cancellation token, provided as a parameter to the task. One major factor here is that the cancellation API has to be compatible with async functions. A `cancel` or `abort` method on the `Promise` prototype is not, whereas cancellation tokens are.

Sketch of public API usage:

``` js
var
  cancel = can.Cancel(),
  xhr    = can.ajax({ url : "...", cancel : cancel );

cancel();
xhr.catch( function( reason ) {
  // reason would be an instance of CancellationError here
});
```

One sound way to implement it is to make the cancellation token itself a promise on which the inner task can be listening and to combine both the task's promise and the cancellation promise using `Promise.race` as the actually output promise of the task.

---

Another benefit of cancellation tokens and the fact that they're externally supplied is that you do not have to worry about propagating a `cancel` or `abort` method all the way up a chain of `then` methods to prevent the cancellation ability from being lost. The caller is responsible to retain a reference to the token.

Directly related to this is the fact that the token is decoupled from the promise, making it easier to provide encapsulation where a consumer only gets access to the promise and does **not** get access to the cancellation.

If you work with a duck-punched method on the `Promise` prototype, you cannot get both. It's an either/or situation: either you always duck-punch the method all the way up the chain to prevent it from being lost, or you never duck-punch the method up the chain to ensure it can be kept private.

**Comment by justinbmeyer** [comment]
For now, I'm alright duck punching to give people some sort of access to cancel the XHR request.  

In future versions of can-connect, we can/should add an optional `cancel` parameter to to the instance and data CRUD methods: `getList`, `get`, `save`, `destroy`, `getListData`, `updateData`, etc.

I'd rather make things possible today instead of waiting to make them perfect.  If we add duck punching, we can make sure there's a `@deprecated` in the docs, and create an issue for the `cancel` parameter.

**Comment by prashantsharmain** [comment]
Closed via #47
